### PR TITLE
cache query results in the servercontainer

### PR DESCRIPTION
**Pull request by icewind1991**
Pimple's `query` has a non negligible cost for repeated querying of the same object (see prs like #21400)

comparisons
- Propfind with single folder result: [-6%](https://blackfire.io/profiles/compare/219bf968-1832-44b4-9da5-f904d2f73713/graph)
- Propfind with ~1k entries result: [-5%](https://blackfire.io/profiles/compare/342b029c-3818-4133-8cba-4ee23c52bf75/graph)

Supersedes #21399 and #21400

cc @DeepDiver1975

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Our simple container already caches the objects but wrapped within a function https://github.com/owncloud/core/blob/453e1bf66e7edb6e2b3d25a093bf2c7662f209a7/lib/private/appframework/utility/simplecontainer.php#L105-L107

The approach in this branch works but I would like to see this inside the simple container so that also the apps container benefits from this.

**Comment by MorrisJobke** [comment]
cc @BernhardPosselt

**Comment by BernhardPosselt** [comment]
Can this be solved upstream?

**Comment by MorrisJobke** [comment]
> Can this be solved upstream?

It is already solved in the SimpleContainer (see my link above) but it seems that this way (object wrapped within a function) isn't that performant.

**Comment by BernhardPosselt** [comment]
Ah ok :)

**Comment by MorrisJobke** [comment]
:+1: from my side

File: lib/private/servercontainer.php [hunk 1]
```diff
@@ -43,6 +45,11 @@ public function __construct() {
 		$this->appContainers = [];
 	}
 
+	public function registerService($name, \Closure $closure, $shared = true) {
```
Comment (nickvergessen) [inline lib/private/servercontainer.php]: What's shared exactly for, is your cache killing it?

Comment (BernhardPosselt) [inline lib/private/servercontainer.php]: Haha, right good point. This PR essentially kills off non shared instances by caching everything.

Comment (MorrisJobke) [inline lib/private/servercontainer.php]: What do you want to say?

Comment (BernhardPosselt) [inline lib/private/servercontainer.php]: Look at the query method below. The method always caches instances so new ones will never be created (which it should when $shared is set to false)

Comment (nickvergessen) [inline lib/private/servercontainer.php]: shared = false means a new instance is returned all the time.
The feature is killed by the caching.
Caching should only be used for services which are shared = true

**Comment by nickvergessen** [comment]
:-1: for now, to make sure no one merges this

**Comment by rullzer** [comment]
Generic simple container solution in https://github.com/owncloud/core/pull/21560

**Comment by nickvergessen** [comment]
Closing as per discussion in https://github.com/owncloud/core/pull/21560#issuecomment-170270541

**Pull request closed without merging**